### DISCUSSION TITLE
consult-xref-history

### DISCUSSION THREAD

**Discussion by jdtsmith** [minad/consult#1039] [Ideas]
I saw this implemented somewhere and can't find it again.  I always get lost in the xref mark stack as you jump to various definitions.  It would be great to have a `M-x consult-xref-history` that presented the xref history stack and where you are within it, similar to consult-mark.  Note that the `xref--history` is a cons of  two stacks, backwards and forwards, and can be global or window-local.

**Comment by minad** [comment]
This should be simple to add. Both consult-mark and consult-global-mark can take a list of markers. Would it be sufficient to wrap one of those?

**Reply by jdtsmith** [reply]
I think the only difference would be indicating where you are in the stack.

**Comment by hmelman** [comment]
Probably this https://takeonrules.com/2024/06/08/adding-a-consult-function-for-visualizing-xref/

**Reply by jdtsmith** [reply]
That's the one! Thanks.

**Comment by jdtsmith** [comment]
I found myself again wishing for a `consult-xref-history`.  This works "OK":

```elisp
(defvar consult--xref-history nil
  "History for the `consult-recent-xref' results.")

(defun consult-recent-xref (&optional markers)
  "Jump to a marker in MARKERS list (defaults to `xref--history'.

The command supports preview of the currently selected marker position.
The symbol at point is added to the future history."
  (interactive)
  (consult--read
   (consult--global-mark-candidates
    (or markers (flatten-list xref--history)))
   :prompt "Go to Xref: "
   ;; :annotate (consult--line-prefix)  ;; broken
   :category 'consult-location
   :sort nil
   :require-match t
   :lookup #'consult--lookup-location
   :history '(:input consult--xref-history)
   :add-history (thing-at-point 'symbol)
   :state (consult--jump-state)))
```

But it has a few flaws.  For some reason, random individual characters are pre-highlighted, even when nothing is entered at the prompt:

<img width="430" height="119" alt="image" src="https://github.com/user-attachments/assets/5e34baba-f512-4bdc-90a1-280abae7a1a0" />

Also, it would be better if the initially selected marker were the _current_ position in xref history not the most recent older one.  Right now invoking `consult-recent-xref` jumps you immediately to one position prior in history.  

The "current" position can be determined by the contents of `xref--history`, which is a compound list, à la:

```
(BACKWARD-STACK . FORWARD-STACK) of markers to visited Xref locations.
```

I.e. if `BACKWARD-STACK` is `nil`, you are at the (oldest) beginning of history.  If `FORWARD-STACK` is `nil` you are at the (newest) end of xref history.  Note that the current marker-position is popped off and pushed upon each `xref` navigation.  

So I tried something like:

```elisp
(defun consult-recent-xref (&optional markers)
  "Jump to a marker in MARKERS list (defaults to `xref--history'.

The command supports preview of the currently selected marker position.
The symbol at point is added to the future history."
  (interactive)
  (let* ((history (xref--get-history))
         (mrk (list (point-marker)))
         (flat (append (reverse (car history)) mrk (cdr history))))
   (consult--read
    (consult--global-mark-candidates (or markers flat))
    :prompt "Go to Xref: "
    ;; :annotate (consult--line-prefix)
    :category 'consult-location
    :default (car (consult--global-mark-candidates mrk))
    :sort nil
    :require-match t
    :lookup #'consult--lookup-location
    :history '(:input consult--xref-history)
    :add-history (thing-at-point 'symbol)
    :state (consult--jump-state))))
```

But this moves the "default selection" to the top (in vertico), where I'd prefer to keep it sorted as-is and have the currently selected candidate fall in its natural location.  Is there a way to tell `consult` to lay out a series of candidates and pre-select a specific one, _in place_?  BTW, a `:default-index` would be nice to have so you don't have to format your chosen candidate.

But then even with the default candidate selection solved, `consult` does not maintain the `xref--history` in the format above, it just blindly visits the markers.  So perhaps a custom state function is what is needed.  It could update the `xref--history` list upon successful return to look again like `(BACKWARD-STACK . FORWARD-STACK)`.

**Reply by oantolin** [reply]
> For some reason, random individual characters are pre-highlighted, even when nothing is entered at the prompt

Don't those indicate which point position is the one in the xref history? I think the history is of specific point locations, not merely lines.

**Reply by jdtsmith** [reply]
Oh good point, that must be right.  If there's a way to select candidate N in position N I can look at implementing a state tracker to respect the backward/forward `xref--history` correctly.  Though maybe "history with a known current position" is a generic enough concept to deserve a special state function, e.g. for `winner-undo/redo`?
